README.md
# ColorWithHSL-iOS

[![Version](https://img.shields.io/cocoapods/v/ColorWithHSL.svg?style=flat-square)](http://cocoapods.org/pods/ColorWithHSL)
[![License](https://img.shields.io/cocoapods/l/ColorWithHSL.svg?style=flat-square)](http://cocoapods.org/pods/ColorWithHSL)
[![Platform](https://img.shields.io/cocoapods/p/ColorWithHSL.svg?style=flat-square)](http://cocoapods.org/pods/ColorWithHSL)
[![CocoaPods](https://img.shields.io/cocoapods/metrics/doc-percent/ColorWithHSL.svg?style=flat-square)](http://cocoapods.org/pods/ColorWithHSL)
[![Build Status](https://img.shields.io/travis/GabrielMassana/ColorWithHSL-iOS/master.svg?style=flat-square)](https://travis-ci.org/GabrielMassana/ColorWithHSL-iOS)

##   What is it?

Swift Extension to convert a HSL colour values into a UIColor Objects.

HSL stands for hue, saturation, and lightness, and is also often called HLS.

Formula from [Wikipedia](https://en.wikipedia.org/wiki/HSL_and_HSV#From_HSL).

## Installation

#### Podfile

```ruby
platform :ios, '8.0'
use_frameworks!

pod 'ColorWithHSL', '~> 0.0'
```

Then, run the following command:

```bash
$ pod install
```

#### Old school

Drag into your project the folder `/ColorWithHSL-iOS`. That's all.

## Example Swift

```swift
        let black: UIColor = UIColor.colorWithHSL(hue: 0.0, saturation: 0.0, lightness: 0.0)
        let white: UIColor = UIColor.colorWithHSL(hue: 0.0, saturation: 0.0, lightness: 1.0)
        let red: UIColor = UIColor.colorWithHSL(hue: 0.0, saturation: 1.0, lightness: 0.5)
        let green: UIColor = UIColor.colorWithHSL(hue: 120.0, saturation: 1.0, lightness: 0.5)
        let blue: UIColor = UIColor.colorWithHSL(hue: 240.0, saturation: 1.0, lightness: 0.5)
```
## Example Objective-C

```objc
#import "ColorWithHSL-Swift.h"

...
    UIColor *black = [UIColor hsl_colorWithHue:0.0 saturation:0.0 lightness:0.0];
    UIColor *white = [UIColor hsl_colorWithHue:0.0 saturation:0.0 lightness:1.0];
    UIColor *red = [UIColor hsl_colorWithHue:0.0 saturation:1.0 lightness:0.5];
    UIColor *green = [UIColor hsl_colorWithHue:120.0 saturation:1.0 lightness:0.5];
    UIColor *blue = [UIColor hsl_colorWithHue:240.0 saturation:1.0 lightness:0.5];
    
        
```
## License

ColorWithHSL-iOS is released under the MIT license. Please see the file called LICENSE.

## Versions

```bash
$ git tag -a 1.0.0 -m 'Version 1.0.0'

$ git push --tags
```

## Author

Gabriel Massana

##Found an issue?

Please open a [new Issue here](https://github.com/GabrielMassana/ColorWithHSL-iOS/issues/new) if you run into a problem specific to ColorWithHSL-iOS, have a feature request, or want to share a comment.


CHANGELOG.md
#### 1.0.6

 - First release
 - Extension will create a UIColor object based on given HSL values.
    - Hue value must be between 0 and 360.
    - Saturation value must be between 0 and 1.
    - Lightness value must be between 0 and 1.
 - Added some Tests




ColorWithHSL-iOS/UIColor+HSL.swift
//
//  UIColor+HSL.swift
//  ColorWithHSL
//
//  Created by GabrielMassana on 02/04/2016.
//  Copyright © 2016 GabrielMassana. All rights reserved.
//
//  https://en.wikipedia.org/wiki/HSL_and_HSV#From_HSL
//

import UIKit

public extension UIColor {
    
    //MARK: - Public method
    
    /**
    Creates UIColor object based on given HSL values.
    
    - parameter hue: CGFloat with the hue value. Hue value must be between 0 and 360.
    - parameter saturation: CGFloat with the saturation value. Saturation value must be between 0 and 1.
    - parameter lightness: CGFloat with the lightness value. Lightness value must be between 0 and 1.
    
    - returns: A UIColor from the given HSL values.
    */
    @objc(hsl_colorWithHue:saturation:lightness:)
    public class func colorWithHSL(hue hue: CGFloat, saturation: CGFloat, lightness: CGFloat) -> UIColor? {
        
        var r: CGFloat = 0
        var g: CGFloat = 0
        var b: CGFloat = 0
        
        guard hue <= 360 && hue >= 0.0 else { return nil }
        guard saturation <= 1.0 && saturation >= 0.0 else { return nil }
        guard lightness <= 1.0 && lightness >= 0.0 else { return nil }
        
        let chroma: CGFloat = (1 - abs((2 * lightness) - 1)) * saturation
        let h60: CGFloat = hue / 60.0
        let x: CGFloat = chroma * (1 - abs((h60 % 2) - 1))
        
        if (h60 < 1) {
            
            r = chroma
            g = x
        }
        else if (h60 < 2)
        {
            r = x
            g = chroma
        }
        else if (h60 < 3)
        {
            g = chroma
            b = x
        }
        else if (h60 < 4)
        {
            g = x
            b = chroma
        }
        else if (h60 < 5)
        {
            r = x
            b = chroma
        }
        else if (h60 < 6)
        {
            r = chroma
            b = x
        }
        
        let m: CGFloat = lightness - (chroma / 2)
        
        r = r + m
        g = g + m
        b = b + m
        
        return UIColor(red: r, green: g, blue: b, alpha: 1.0)
    }
}

ColorWithHSL/Example/AppDelegate.swift
//
//  AppDelegate.swift
//  ColorWithHSL
//
//  Created by GabrielMassana on 02/04/2016.
//  Copyright © 2016 GabrielMassana. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



ColorWithHSLTests/ColorWithHSLTests.swift
//
//  ColorWithHSLTests.swift
//  ColorWithHSLTests
//
//  Created by GabrielMassana on 02/04/2016.
//  Copyright © 2016 GabrielMassana. All rights reserved.
//

import XCTest

@testable import ColorWithHSL

class ColorWithHSLTests: XCTestCase {
    
    override func setUp() {
        
        super.setUp()
    }
    
    override func tearDown() {

        super.tearDown()
    }
    
    //MARK: - Valid

    func test_colorWithHSL_newObjectReturned_minimum() {
        
        let color = UIColor.colorWithHSL(hue: 0.0, saturation: 0.0, lightness: 0.0)
        
        XCTAssertNotNil(color, "A valid Color object wasn't created");
    }

    func test_colorWithHSL_newObjectReturned_maximum() {
        
        let color = UIColor.colorWithHSL(hue: 360.0, saturation: 1.0, lightness: 1.0)
        
        XCTAssertNotNil(color, "A valid Color object wasn't created");
    }
    
    //MARK: - NoValid
    
    func test_colorWithHSL_outOfRangeHue_over() {
        
        let color = UIColor.colorWithHSL(hue: 361.0, saturation: 0.5, lightness: 0.5)
        
        XCTAssertNil(color, "A valid Color object was created");
    }

    func test_colorWithHSL_outOfRangeHue_under() {
        
        let color = UIColor.colorWithHSL(hue: -1.0, saturation: 0.5, lightness: 0.5)
        
        XCTAssertNil(color, "A valid Color object was created");
    }
    
    func test_colorWithHSL_outOfRangeSaturation_over() {
        
        let color = UIColor.colorWithHSL(hue: 180.0, saturation: 1.1, lightness: 0.5)
        
        XCTAssertNil(color, "A valid Color object was created");
    }
    
    func test_colorWithHSL_outOfRangeSaturation_under() {
        
        let color = UIColor.colorWithHSL(hue: 180.0, saturation: -0.1, lightness: 0.5)
        
        XCTAssertNil(color, "A valid Color object was created");
    }
    
    func test_colorWithHSL_outOfRangeLightness_over() {
        
        let color = UIColor.colorWithHSL(hue: 180.0, saturation: 0.5, lightness: 1.1)
        
        XCTAssertNil(color, "A valid Color object was created");
    }
    
    func test_colorWithHSL_outOfRangeLightness_under() {
        
        let color = UIColor.colorWithHSL(hue: 180.0, saturation: 0.5, lightness: -0.1)
        
        XCTAssertNil(color, "A valid Color object was created");
    }
    
    //MARK: - SpecificColor
    
    func test_colorWithHSL_red() {
        
        let redColor = UIColor.colorWithHSL(hue: 0.0, saturation: 1.0, lightness: 0.5)
        
        XCTAssertEqual(redColor, UIColor.redColor(), "A red Color object wasn't created");
    }
    
    func test_colorWithHSL_green() {
        
        let greenColor = UIColor.colorWithHSL(hue: 120.0, saturation: 1.0, lightness: 0.5)
        
        XCTAssertEqual(greenColor, UIColor.greenColor(), "A green Color object wasn't created");
    }
    
    func test_colorWithHSL_blue() {
        
        let blueColor = UIColor.colorWithHSL(hue: 240.0, saturation: 1.0, lightness: 0.5)
        
        XCTAssertEqual(blueColor, UIColor.blueColor(), "A blue Color object wasn't created");
    }
}


ColorWithHSL/Example/ViewController.swift
//
//  ViewController.swift
//  ColorWithHSL
//
//  Created by GabrielMassana on 02/04/2016.
//  Copyright © 2016 GabrielMassana. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        
        super.viewDidLoad()

        let black: UIColor = UIColor.colorWithHSL(hue: 0, saturation: 0, lightness: 0)!
        let white: UIColor = UIColor.colorWithHSL(hue: 0, saturation: 0, lightness: 1)!
        let red: UIColor = UIColor.colorWithHSL(hue: 0, saturation: 1, lightness: 0.5)!
        let green: UIColor = UIColor.colorWithHSL(hue: 120, saturation: 1, lightness: 0.5)!
        let blue: UIColor = UIColor.colorWithHSL(hue: 240, saturation: 1, lightness: 0.5)!
        
        print("black", black)
        print("white", white)
        print("red", red)
        print("green", green)
        print("blue", blue)
    }
}

